File: Cursor.cs
using System.Linq;
using System.Text;

public partial record Cursor
{
    public static Cursor Empty = new Cursor(string.Empty, new Context(new Operators()));

    public int Line { get; init; } = 1;

    public int Col { get; init; } = 1;
    
    public Cursor(string text, Context context)
    : this(text.EnumerateRunes().ToList(), 0, context)
    {}

    public Cursor Next()
    {
        if (!More) 
            return this;

        if (Current.ToString() == "\n")
            return new Cursor(Source, Offset + 1, Context) { Line = Line + 1, Col = 1 };

        if (Current.ToString() != "\r")
            return new Cursor(Source, Offset + Current.Utf16SequenceLength, Context) { Line = Line, Col = Col + 1 };

        var crlf = Offset + 1 < Source.Count && Source[Offset + 1].ToString() == "\n";
        return new Cursor(Source, Offset + (crlf ? 2 : 1), Context) { Line = Line + 1, Col = 1 };
    }
}

public sealed partial record TokenCursor
{
    public TokenCursor Next() => More ? new TokenCursor(this, Offset + 1) : this;    

    public int Line => Current.Start.Line;

    public int Col => Current.Start.Col;
}

internal static class RuneExtensions
{
    public static bool Is(this Rune rune, string comparand) =>
        rune.ToString() == comparand;

    public static bool IsNot(this Rune rune, string comparand) =>
        rune.ToString() != comparand;
}

File: Cursor.fs
module Cursor

open System.Text

type Cursor =
    { line: int
      pos: int
      index: int
      source: Rune array
    }

    member c.More = c.index < c.source.Length

    member c.Current = if c.More then c.source[c.index] else new Rune(0)

    member c.Str = c.Current.ToString()

    member c.Next = 
        if c.More then 
            match c.Str with
            | "\n" ->
                { c with index = c.index + 1; line = c.line + 1; pos = 1 }
            | "\r" when c.index > c.source.Length - 2 || c.source[c.index + 1].ToString() <> "\n" -> 
                { c with index = c.index + 1; line = c.line + 1; pos = 1 }
            | _ ->
                { c with index = c.index + 1; pos = c.pos + 1 }
        else c

type Cspan = Cursor * Cursor

let spanned (span: Cspan) =
    let (start, next) = span

    [start.index .. next.index - 1]
    |> List.map (fun i -> start.source[i].ToString())
    |> String.concat ""

let makeCursor (source:string) =
    let runes = source.EnumerateRunes() |> Array.ofSeq
    { source = runes; index = 0; line = 1; pos = 1 }

let rec getCursors' c runes =
    seq {
        let c2 =
            match runes with
            | [] -> c
            | r::n::rs when r.ToString() = "\r" && n.ToString() <> "\n" ->
                {c with index = c.index + 1; line = c.line + 1; pos = 1}
            | r::rs when r.ToString() = "\n" || r.ToString() = "\r" ->
                {c with index = c.index + 1; line = c.line + 1; pos = 1}
            | r::rs ->
                {c with index = c.index + 1; pos = c.pos + 1 }
        match runes with
        | [] ->
            ignore ()
        | _::rs ->
            yield c
            yield! (getCursors' c2 rs)
    }

let getCursors runes =
    getCursors'
        { source = runes |> Array.ofList; index = 0; line = 1; pos = 1 }
        runes
